workflow/metrics/controller_metrics.go

```diff
@@ -48,24 +24,72 @@ func (wc *controllerCollector) collectWorkflowStatuses(wfs []*wfv1.Workflow) []p
 			ConstLabels: map[string]string{"status": string(phase)},
 		}
 	}
-	gauges := map[wfv1.NodePhase]prometheus.Gauge{
-		wfv1.NodePending:   prometheus.NewGauge(getOptsByPahse(wfv1.NodePending)),
-		wfv1.NodeRunning:   prometheus.NewGauge(getOptsByPahse(wfv1.NodeRunning)),
-		wfv1.NodeSucceeded: prometheus.NewGauge(getOptsByPahse(wfv1.NodeSucceeded)),
-		wfv1.NodeSkipped:   prometheus.NewGauge(getOptsByPahse(wfv1.NodeSkipped)),
-		wfv1.NodeFailed:    prometheus.NewGauge(getOptsByPahse(wfv1.NodeFailed)),
-		wfv1.NodeError:     prometheus.NewGauge(getOptsByPahse(wfv1.NodeError)),
+
+	var controllerCollector controllerCollector
+
+	controllerCollector.informer = informer
+	controllerCollector.addWorkflowInformerHandler()
+
+	controllerCollector.gauges = map[wfv1.NodePhase]prometheus.Gauge{
+		wfv1.NodePending:   prometheus.NewGauge(getOptsByPhase(wfv1.NodePending)),
+		wfv1.NodeRunning:   prometheus.NewGauge(getOptsByPhase(wfv1.NodeRunning)),
+		wfv1.NodeSucceeded: prometheus.NewGauge(getOptsByPhase(wfv1.NodeSucceeded)),
+		wfv1.NodeSkipped:   prometheus.NewGauge(getOptsByPhase(wfv1.NodeSkipped)),
+		wfv1.NodeFailed:    prometheus.NewGauge(getOptsByPhase(wfv1.NodeFailed)),
+		wfv1.NodeError:     prometheus.NewGauge(getOptsByPhase(wfv1.NodeError)),
+	}
+
+	return &controllerCollector
+}
+
+func (cc *controllerCollector) addWorkflowInformerHandler() {
+	cc.informer.AddEventHandler(
```

Comment:
This is an interesting concept - but I raise the following problem:- the metrics and controller code in very much intertwined with one another. This PR in its current state doubles-down on that complexity by adding an informer to the metrics.

We can make this code easier understand and easier to test (I can see test that cannot be mocked in the controller that could be). 

We can fix this easily but make the following delineation responsibilities:

* The `metrics` package is only responsible for tracking, TTL, and reporting the metrics.
* The `controller` package is only responsible for telling the metrics about events it would find interesting

We can do this by inverting the relationship and making the `WorkflowController` only responsible for telling the interface when things happen. 

I'll link some code shortly.

Reply:
I see what you're saying. It should be simple to add this to the Workflow informer on the controller. We can then communicate with the metrics "service" to actually emit the metrics. 

Seems like it would be good to merge https://github.com/argoproj/argo/issues/2927 into this.

Reply:
I've dug a bit deeper and found out that we now have 3 different workflow informers. I think we need to collapse them into a single informer which multiple handlers. I've been sketching this out here:

https://github.com/argoproj/argo/pull/2921/files#diff-0514b59bc1689e4e41d87880fcbaae58R12

Reply:
Can we delete the legacy metrics?